src/quacc/recipes/espresso/phonons.py

```diff
@@ -45,21 +72,225 @@ def phonon_job(
+        - name: "ph_test_job"
+        - job: [quacc.recipes.espresso.phonons.phonon_job][]
+
+    3. (n * m) / nblocks ph.x calculations
+        - name: "ph_job"
+        - job: [quacc.recipes.espresso.phonons.phonon_job][]
+
+    4. ph.x calculation to gather data and diagonalize each dynamical matrix
+        - name: "ph_recover_job"
+        - job: [quacc.recipes.espresso.phonons.phonon_job][]
+
+    Parameters
+    ----------
+    atoms
+        Atoms object
+    nblocks
+        The number of representations to group together in a single job.
+        This will reduce the amount of data produced by a factor of nblocks.
+        If nblocks = 0, each job will contain all the representations for a
+        single q-point.
+    job_params
+        Custom parameters to pass to each Job in the Flow. This is a dictinoary where
+        the keys are the names of the jobs and the values are dictionaries of parameters.
+    job_decorators
+        Custom decorators to apply to each Job in the Flow. This is a dictionary where
+        the keys are the names of the jobs and the values are decorators.
+
+    Returns
+    -------
+    RunSchema
+        Dictionary of results from [quacc.schemas.ase.summarize_run][]
+    """
+
+    calc_defaults = {
+        "relax_job": {
+            "input_data": {
+                "control": {"forc_conv_thr": 5.0e-5},
+                "electrons": {"conv_thr": 1e-12},
+            }
+        },
+        "ph_init_job": {
+            "input_data": {
+                "inputph": {
+                    "lqdir": True,
+                    "tr2_ph": 1e-12,
+                    "only_init": True,
+                    "verbosity": "high",
+                }
+            }
+        },
+        "ph_job": {
+            "input_data": {
+                "inputph": {
+                    "tr2_ph": 1e-12,
+                    "alpha_mix(1)": 0.1,
+                    "verbosity": "high",
+                }
+            }
+        },
+        "ph_recover_job": {
+            "input_data": {
+                "inputph": {"recover": True, "lqdir": True, "verbosity": "high"}
+            }
+        },
+    }
+
+    job_params = recursive_dict_merge(calc_defaults, job_params)
+    pw_job, ph_init_job, ph_job, recover_ph_job = customize_funcs(
+        ["relax_job", "ph_init_job", "ph_job", "ph_recover_job"],
+        [relax_job, phonon_job, phonon_job, phonon_job],
+        parameters=job_params,
+        decorators=job_decorators,
+    )
+
+    ph_job_input_data = job_params.get("ph_job", {}).get("input_data", {})
+
+    pw_job_results = pw_job(atoms)
+
+    ph_test_job_results = ph_init_job(
+        pw_job_results["dir_name"], input_data=ph_job_input_data
+    )
+
+    grid_results = _grid_phonon_subflow(
+        ph_job_input_data, ph_test_job_results, ph_job, nblocks=nblocks
+    )
+
+    prev_dirs = {}
+
+    for result in grid_results:
+        prev_dirs[result["dir_name"]] = ["**/dynmat.*.*.xml", "**/tensors.xml"]
```

Comment:
While I have not tested it yet, I imagine that this will not work. You are trying to populate a dictionary with a future value that is not known until runtime. In other words, the key `results["dir_name"]` is not known _a priori_ and will not behave as expected. Assignments like this should not be done within a `@flow`.

The previous scenario was:

```python
copy_back = [result["dir_name"] for result in grid_results]
ph_job_input_data["recover"] = True
return recover_ph_job(copy_back, input_data=ph_job_input_data)
```

This worked because `copy_back` contains a `list` of future objects. No operations are done on these future objects; rather, they are passed to `recover_ph_job`, which then knows to resolve the list of futures before starting that calculation. That is about the extent of things you can do with futures in a `@flow`.

Reply:
(I have seen your edit) I believed this would work because it worked with list... I thought they are both getter/setter under the hood and I would not make any difference.

Would some kind of deepcopy or something else would make sure that it always work?

Reply:
The deepcopy won't be relevant here. Basically, the way things work in `@flow`s is that every return object from a `@job` is a `Future` (or equivalent), and you can't do any* operations on a `Future` other than pass it as the input to another `@job`. When it's passed to the `@job`, it will be auto-resolved inside that `@job`. The one exception is, generally, a `__getitem__` call where doing `Future["key"]` will basically defer the evaluation of `__getitem__(self, key)` until the `Future` is in fact resolved, e.g. something we often do where we pass `results["atoms"]` as the output of one `@job` to the input of another. It knows not to index it until later.

Before, you were appending `Future`s to a list (this is fine; the operation doesn't care what the value of `Future` will be) and passing that `list[Future]` to a `@job`. All okay. Here, you are effectively doing `dict[Future] = list[str]`. The key is not a `str` but rather a `Future`, so it's unclear to me in advance what will happen. I don't know off-hand how the auto-resolving of `Future`s happens across the various workflow engines.

This didn't work before because you were effectively doing `Future[str] = bool`, and you can't assign a property to a `Future`.

```python
input_data = grid_results[-1]["parameters"]["input_data"]
input_data["inputph"]["recover"] = True
```

In this commit, you're assigning a property to a dictionary whose key is a `Future`, and my brain hurts thinking about it so tl;dr --- there will only one way to find out...